### PR TITLE
add graceful mode that does not throw exceptions by default (FF-953)

### PR DESCRIPTION
**Pull request by leoromanovsky**

### CODE REVIEW

File: src/EppoClient.php [hunk 1]
```diff
@@ -226,7 +285,7 @@ private function getAssignmentVariation(string $subjectKey, string $flagKey, arr
         }
 
         $overrideVariation = $this->getSubjectOverrideVariation($subjectKey, $experimentConfig);
-        
+
```
Comment (leoromanovsky) [inline src/EppoClient.php:288]: I enabled a linter in vscode and it removed some whitespace

File: tests/EppoClientTest.php [hunk 2]
```diff
@@ -267,11 +265,37 @@ public function testHandlesLoggingException()
...
+        $mockConfigRequester = $this->getExperimentConfigurationRequesterMock(self::MOCK_EXPERIMENT_CONFIG, new Exception('config requester error'));
+        $mockLogger = $this->getMockBuilder(LoggerInterface::class)->getMock();
+
+        $subjectAttributes = [['foo' => 3]];
+        $client = EppoClient::createTestClient($mockConfigRequester, $pollerMock, $mockLogger, false);
+
+        $this->expectException(Exception::class);
+        $client->getAssignment('subject-10', self::EXPERIMENT_NAME, $subjectAttributes);
+    }
+
```
Comment (leoromanovsky) [inline tests/EppoClientTest.php:293]: new tests - do you want to see these extended to all the public methods?

File: src/EppoClient.php [hunk 3]
```diff
@@ -42,6 +42,9 @@ class EppoClient
     /** @var PollerInterface */
     private $poller;
 
+    /** @var bool */
+    private $isGracefulMode;
```
Comment (aarsilv) [inline src/EppoClient.php:46]: consider naming something more explicit like `isGracefulHandleExceptions`

File: src/EppoClient.php [hunk 4]
```diff
@@ -117,9 +123,18 @@ public static function getInstance(): EppoClient
...
-        $assignmentVariation = $this->getAssignmentVariation($subjectKey, $flagKey, $subjectAttributes, self::VARIANT_TYPE_STRING);
-        return  $assignmentVariation ? strval($assignmentVariation->typedValue) : null;
+    public function getStringAssignment(string $subjectKey, string $flagKey, array $subjectAttributes = []): ?string
+    {
+        try {
+            $assignmentVariation = $this->getAssignmentVariation($subjectKey, $flagKey, $subjectAttributes, self::VARIANT_TYPE_STRING);
+            return  $assignmentVariation ? strval($assignmentVariation->typedValue) : null;
+        } catch (Exception $exception) {
+            if ($this->isGracefulMode) {
+                error_log('[Eppo SDK] Error getting string assignment: ' . $exception->getMessage());
```
Comment (aarsilv) [inline src/EppoClient.php:133]: yay logging 👏

File: src/EppoClient.php [hunk 5]
```diff
@@ -117,9 +123,18 @@ public static function getInstance(): EppoClient
...
+                return null;
+            }
+            throw $exception;
```
Comment (aarsilv) [inline src/EppoClient.php:136]: Like we did in the JS client, this code can be deduped into a private helper function

File: tests/EppoClientTest.php [hunk 6]
```diff
@@ -267,11 +265,37 @@ public function testHandlesLoggingException()
         $this->assertEquals('control', $assignment);
     }
 
+    public function testGracefulModeDoesNotThrow()
```
Comment (aarsilv) [inline tests/EppoClientTest.php:268]: 🙌

File: tests/EppoClientTest.php [hunk 7]
```diff
@@ -267,11 +265,37 @@ public function testHandlesLoggingException()
...
+        $mockLogger = $this->getMockBuilder(LoggerInterface::class)->getMock();
+
+        $subjectAttributes = [['foo' => 3]];
+        $client = EppoClient::createTestClient($mockConfigRequester, $pollerMock, $mockLogger, true);
+
+        $assignment = $client->getAssignment('subject-10', self::EXPERIMENT_NAME, $subjectAttributes);
+        $this->assertNull($assignment);
+    }
+
+    public function testNoGracefulModeThrows()
```
Comment (aarsilv) [inline tests/EppoClientTest.php:281]: 🙌

File: tests/EppoClientTest.php (hunk 2 above)
Comment (aarsilv) [reply tests/EppoClientTest.php:293]: I think it would be worth it to be thorough as long as it could be done in a way that reused 95% of the code

**Review by aarsilv — APPROVED** [review]
Approving with some recommended tweaks in the comments

File: src/EppoClient.php (hunk 3 above)
Comment (leoromanovsky) [reply src/EppoClient.php:46]: In the other SDKs are named `isGracefulMode` - I agree you suggestion is more clear. I would like to update them all at the same time to be consistent.

File: src/EppoClient.php [hunk 8]
```diff
@@ -181,9 +223,18 @@ public function getParsedJSONAssignment(string $subjectKey, string $flagKey, arr
      * @throws InvalidArgumentException
      * @throws SimpleCacheInvalidArgumentException
      */
-    public function getJSONStringAssignment(string $subjectKey, string $flagKey, array $subjectAttributes = []): string {
-        $parsedJsonValue = $this->getParsedJSONAssignment($subjectKey, $flagKey, $subjectAttributes);
-        return isset($parsedJsonValue) ? json_encode($parsedJsonValue) : null;
+    public function getJSONStringAssignment(string $subjectKey, string $flagKey, array $subjectAttributes = []): ?string
```
Comment (leoromanovsky) [inline src/EppoClient.php:226]: fyi @aarsilv I found we were returning `string` here; in php returning `null` is not allowed so I changed to `?string` - this also matches the other functions.

bumping to 1.3.0 as its a minor api change.

Comment (aarsilv) [reply src/EppoClient.php:210]: oh good catch!

**Pull request merged**